Feature: pokemon and quest reset on event change

hamster007Github opened the pull request:
# Feature description
1. **Pokemon reset**: MAD don't update pokemon in DB, if pokemon change because of a event start or event end. So this new feature check for events with different pokemon pool. If a relevant event change is detected, pokemon, which are scanned before event change (last_modified) and despawn after event change (disappear_time), will be removed from MAD DB. So MAD will rescan them.
2. **Quest delete**: add quest delete feature, which delete quests from MAD DB on quest related event change.

new plugin.ini parameter - general:
- `reset_pokemons`: set to `true`, to enable feature
- `reset_pokemons_truncate`:  set to `true`, to use SQL TRUNCATE to delete all pokemon DB entries on event change. Recommended for big MAD instances with huge DBs and many devices. Set to `false`, to use SQL DELETE to delete only the event change effected pokemon DB entries.

new plugin.ini parameter - Quest resets:
- `enable_quest_delete`: Whether or not to enable auto Quest delete by SQL TRUNCATE trs_quest table of MAD database
- `delete_quests_for`: Define event types, which triggers quest delete for their start, end or both. see `reschedule_for` for details.

**Remark: not well tested, because of missing event change :-/ Please review and test. I do welcome all kinds of feedback.**
@crhbetz please don't merge until positive feedback

# Change details
1. I split the plugin loop into a fast one (60 s loop) to check for event change and a slow one (include all old functions) by adding a counter
2. I added a _pokemon_events list for pokemon relevant events (Type: community-day, spotlight-hour or entries in spawns)
3. I added a cooldown function for pokemon resets with fix 30 minutes, to limit SQL DELETE request to one per event change
4. Convert event times to UTC time (pokemon table use UTC) by subtract self.tz_offset
5. SQL delete request: `DELETE FROM pokemon WHERE last_modified < *datetime event start/end in UTC* AND disappear_time > *datetime event start/end in UTC*` or `TRUNCATE pokemon`

Commit:
add option for automatically remove pokemon from DB on event start and event end

Commit:
add cooldown function for pokemon reset and some code cleanup

crhbetz commented:
Thanks for taking the time, this would be quite a useful addition I think!

Don't have a lot of time right now to properly review and/or start testing, just noting down a few thoughts from taking a quick look:

- I don't know a lot about databases, but I think a `DELETE FROM pokemon WHERE` query has the potential to cause a huge database lockup on large `pokemon` tables. Maybe there should at least be an option to just `TRUNCATE pokemon` for people who don't care about stats, which I think runs pretty much instantly with no locking up
- `eventscheck_cycle_counter` I guess works but it seems a bit of a weird solution to me. Why not store a `self._last_checked_events` and then run the "old main loop" `if datetime.now() - self._last_checked_events > self.__sleep` ?

Commit:
include feedback: use date instead of counter for event check loop, add plugin.ini option reset_pokemons_truncateto use TRUNCATE instead of DELETE

hamster007Github commented:
Thank you for your feedback. Fixed your findings.

Commit:
fixed self.tz_offset calculation in case local and UTC time are on different days\nfixed wrong logger variable in _check_pokemon_resets(), which abort pokemon reset\nreduced timewindow for eventcheck to prevent multiple TRUNCATE, if MAD is restarted just after event change

hamster007Github commented:
Fixed a bug in the old existing code and in new feature code. Tested with manual added event. Seems to work

Commit:
fixed SQL query execution (commit=true and using strings for datetime)

hamster007Github commented:
After last commit, pokemon reset feature is working fine with using DELETE (reset_pokemons_truncate = false) on end of spotlight-hour on a live MAD system.

Commit:
removed non ANSI chars from plugin.ini.example

Commit:
simplify check to remove cooldown function

Commit:
rename feature of 'quest reset' to 'quest reschedule' and add new quest reset feature, which delete quests on event change. Both methods can be used in parallel. README adapted.

hamster007Github committed:
Merge branch 'master' into pokemon_reset

hamster007Github commented:
Added additional feature to reset quests (delete quests in MAD database) on quest related event change. Also requested in [Issue: Allow an option to empty the database after a quest reset](https://github.com/ccev/mp-eventwatcher/issues/3). 
Adapted README with new feature.
Remark: renamed the old quest reset plugin.ini parameter to clearly separate the two methods. See also updated PR description.

ccev commented on autoevents.py:
```diff
@@ -138,13 +151,76 @@ def _convert_time(self, time_string, local=True):
+                    self._mad["mapping_manager"].update()
+                    break
+            # event end during last check?
+            if "end" in self.__quests_reset_types.get(event["type"], []):
```
code duplication:

please change the the if statement to something like this:

```py
to_check = self.__quests_reset_types.get(event["type"], [])
if "start" in to_check or "end" in to_check:
```

hamster007Github replied:
corrected by similar code as you used in quest reschedule function. See https://github.com/ccev/mp-eventwatcher/pull/8/commits/eb4d253da60e39a7864ca7ba52bb7bab3860ff67

ccev commented on autoevents.py:
```diff
@@ -138,13 +151,76 @@ def _convert_time(self, time_string, local=True):
+                self._reset_pokemon(event["start"] - timedelta(hours=self.tz_offset))
+                break
+            # event end during last check?
+            if self._last_pokemon_reset_check < event["end"] <= now:
```
see below comment. same applies here

hamster007Github replied:
duplication remove. See https://github.com/ccev/mp-eventwatcher/pull/8/commits/eb4d253da60e39a7864ca7ba52bb7bab3860ff67

Commit:
fix README

ccev commented:
Hey, only just seeing the PR. Thanks a ton!!

A couple notes after glancing over your code changes:

not a big fan of the pokemon trunacte/delete feature. as far as i can tell there are 2 key things missing:
- you should take care of weather changes
- you should just update the db entries instead of deleting them

imo, this is best done in a separate plugin to eventwatcher. there you could try to make your plugin overwrite the get_latest_mitm method as well as all DbPogoProtoSubmit methods that insert mons. just have this reset every full hour and it should be fine.

not a big fan of renaming existing config options either. this will just make things unnecessarily complicated for everyone 🙂 (though i agree the existing config option names are a bit amibgious). Instead you should put your new quest reset approach in a new config category and name it something like `Quest delete`. 

maybe, instead of deleting quests you could set their quest_timestamps to something from the past? so if there's a false-positive, mappers can just reverse this change instead of having to rescan? Not sure if this makes much sense, just an idea

Commit:
removed duplicated code

Commit:
reverted parameter renaming for existing parameters, using name 'quest delete' for new quest reset feature

hamster007Github commented:
Thank you for your feedback 👍
> ...
> not a big fan of the pokemon trunacte/delete feature. as far as i can tell there are 2 key things missing:
>   - you should take care of weather changes
>   - you should just update the db entries instead of deleting them

I'm not planning to create separate plugin or handle weather changes in the next time. Can just provide as it is for now. I tried to find which pokemon values of the SQL entries needs to be modified to allow a rescan of mons, but I was not successful.

> not a big fan of renaming existing config options either. this will just make things unnecessarily complicated for everyone 🙂 (though i agree the existing config option names are a bit amibgious). Instead you should put your new quest reset approach in a new config category and name it something like `Quest delete`.

Reverted renaming with commit https://github.com/ccev/mp-eventwatcher/pull/8/commits/3d0885d2c864010c64a383e8e6d99f783ef1cf57

cyphercris commented:
Test pokemon restart with delete option and I have this error now:
`Failed executing query: INSERT INTO pokemon (encounter_id, spawnpoint_id, pokemon_id, latitude, longitude, disappear_time, individual_attack, individual_defense, individual_stamina, move_1, move_2, cp, cp_multiplier, weight, height, gender, catch_prob_1, catch_prob_2, catch_prob_3, rating_attack, rating_defense, weather_boosted_condition, last_modified, costume, form, seen_type) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s) ON DUPLICATE KEY UPDATE last_modified=VALUES(last_modified), disappear_time=VALUES(disappear_time), individual_attack=VALUES(individual_attack), individual_defense=VALUES(individual_defense), individual_stamina=VALUES(individual_stamina), move_1=VALUES(move_1), move_2=VALUES(move_2), cp=VALUES(cp), cp_multiplier=VALUES(cp_multiplier), weight=VALUES(weight), height=VALUES(height), gender=VALUES(gender), catch_prob_1=VALUES(catch_prob_1), catch_prob_2=VALUES(catch_prob_2), catch_prob_3=VALUES(catch_prob_3), rating_attack=VALUES(rating_attack), rating_defense=VALUES(rating_defense), weather_boosted_condition=VALUES(weather_boosted_condition), costume=VALUES(costume), form=VALUES(form), pokemon_id=VALUES(pokemon_id), fort_id=NULL, cell_id=NULL, latitude=VALUES(latitude), longitude=VALUES(longitude), spawnpoint_id=VALUES(spawnpoint_id), seen_type=VALUES(seen_type) ([11830320785734929980, 4870577968565, 682, 53.853879351719186, 22.988945732338166, '2022-02-08 17:52:20', 10, 5, 2, 320, 85, 924, 0.719399094581604, 0.4384497106075287, 0.19451135396957397, 2, 0.13900489, 0.20108414, 0.25868744, None, None, 0, '2022-02-08 17:00:16', 0, 0, 'encounter']), error: 1062 (23000): Duplicate entry '11830320785734929980' for key 'PRIMARY'
`
